CASSANDRA-18618: Separate task for running static analysis

jacek-lewandowski opened the pull request.

jacek-lewandowski commented:
https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch/2517/

jacek-lewandowski commented:
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/744/workflows/45d20631-96a2-4116-9a92-14fadbaaddb8/jobs/15328 (see the "Run static code analysis step")

michaelsembwever commented:
can we please move the different check/list ant targets into a new .build/build-checks.xml files? (or separate .build/build-xxx.xml file)

it really helps readability of build.xml (encapsulating concerns).

michaelsembwever commented on .build/build-rat.xml:
```diff
@@ -106,7 +106,7 @@
     </target>
 
-    <target name="_assert_rat_output">
+    <target name="_assert_rat_output" unless="${rat.skip}">
```
```suggestion
    <target name="_assert_rat_output" unless="rat.skip">
```

?

jacek-lewandowski replied:
what do you mean? it is defined in the same way as in https://github.com/apache/cassandra/pull/2434/files#diff-d274784b2fecdbee879b6175c05a9860374c65c0c0b7e87efe8ec66c0fc3d838R50

michaelsembwever replied:
that is a mistake.  it should just be the string literal.

michaelsembwever commented on .circleci/config.yml:
```diff
@@ -4131,7 +4131,7 @@ jobs:
             # Loop to prevent failure due to maven-ant-tasks not downloading a jar..
             for x in $(seq 1 3); do
-                ${ANT_HOME}/bin/ant realclean; ${ANT_HOME}/bin/ant jar dtest-jar
+                ${ANT_HOME}/bin/ant realclean; ${ANT_HOME}/bin/ant jar dtest-jar -Dno-checkstyle=true -Dno-check=true
```
it shouldn't be necessary to supply both `-Dno-checkstyle=true -Dno-check=true`

i.e. `-Dno-check=true` should imply `-Dno-checkstyle=true -Drat.skip=true -Declipse-warnings.skip=true -Ddependency-check.skip=true`

here's a patch i was playing with before… 
https://github.com/apache/cassandra/compare/trunk...thelastpickle:cassandra:mck/buildfiles-linter-docs/4.1

( that patch also allows you to specify there shorter form of `-Dno-linter=` )

jacek-lewandowski replied:
This is for older branches - are we going to apply those changes to those older branches as well?

michaelsembwever replied:
erh, no. go nuts then.

jacek-lewandowski commented:
@michaelsembwever I've rebased and applied the changes according to what was concluded on the ML

michaelsembwever commented on .circleci/config.yml:
```diff
@@ -4131,7 +4131,7 @@ jobs:
             # Loop to prevent failure due to maven-ant-tasks not downloading a jar..
             for x in $(seq 1 3); do
-                ${ANT_HOME}/bin/ant realclean; ${ANT_HOME}/bin/ant jar dtest-jar
+                ${ANT_HOME}/bin/ant realclean; ${ANT_HOME}/bin/ant jar dtest-jar -Dno-checkstyle=true -Drat.skip=true
```
are these still needed if `check` is now its own target …?

jacek-lewandowski replied:
yes, because dtest jars are built for older Cassandra versions and we decided to not port this change to older branches

michaelsembwever commented on build.xml:
```diff
@@ -540,6 +539,14 @@
+        <antcall target="_assert_rat_output" inheritrefs="true"/>
+        <antcall target="checkstyle" inheritrefs="true"/>
+        <antcall target="checkstyle-test" inheritrefs="true"/>
+        <antcall target="eclipse-warnings" inheritrefs="true"/>
```
what about dependency-check ?

jacek-lewandowski replied:
We should have a dedicated ticket for that. Unless the project dependencies are changed, the developer cannot be blamed for CVEs. 

Also, if the dependency set is not changed, it does not make sense to download and analyze the project against the whole database because we can be affected only by CVEs which are discovered recently.

There are a couple of valuable things we can do about dependency checks, including license analysis, and I'd like to avoid mixing this PR with that.

jacek-lewandowski commented on .build/build-git.xml:
```diff
@@ -51,7 +51,34 @@
+            <arg value="--ignore-all-space"/>
+            <arg value="--merge-base"/>
+            <arg value="${_ref_branch}"/>
+            <arg value="${build.helpers.dir}/cassandra-build-deps-template.xml"/>
```
actually we should not care about cassandra-build-deps

jacek-lewandowski commented on .build/build-owasp.xml:
```diff
@@ -16,9 +16,11 @@
+         xmlns:unless="ant:unless">
     <property name="dependency-check.version" value="8.3.1"/>
-    <property name="dependency-check.home" value="${build.dir}/dependency-check-ant-${dependency-check.version}"/>
+    <property name="dependency-check.home" value="${user.home}/.ant/dependency-check-ant-${dependency-check.version}"/>
```
I set it to user.home because downloading the database is painfully slow and it can be shared between projects / worktrees. Redownloading it after ant realclean was frustrating

michaelsembwever replied:
is it specific to `ant` ?  

could it be under `${user.home}/.owasp/dependency-check-ant-${dependency-check.version}` instead?

how big does it get? should it go under `${user.home}/.cache/owasp/…` instead?  (mine was ~500M , but it will grow over time…)

@smiklosovic thoughts?

jacek-lewandowski replied:
I'm fine with any location under user home. It takes ~250M currently

jacek-lewandowski replied:
@michaelsembwever maybe there is some public server which hosts such database and we could connect to it instead (or at least as a first choice)?

smiklosovic replied:
@michaelsembwever `~/.cache` seems to be a good dir to put it under. We are really caching it and there are other caches in place already on desktops. Maybe `~/.cache/cassandra-owasp` would be the best to say that such owasp cache is from Cassandra project. Otherwise who would know?

smiklosovic replied:
@jacek-lewandowski what happens if `~/.cache` dir does not exist? Highly improbable but still ... does it create parent dirs too?

jacek-lewandowski replied:
yes, it creates the parent directories as well, I've set it to `~/.cache/dependency-check-ant/owasp-db`

michaelsembwever replied:
i agree with @smiklosovic , `~/.cache/cassandra-owasp-dependency-check` is a better name.

Users will want to know where a half gigabyte folder came from.

jacek-lewandowski commented on .build/build-owasp.xml:
```diff
@@ -83,4 +85,67 @@
+        </exec>
+
+        <loadfile property="wrongly-licensed" srcFile="${build.dir}/owasp/license-report.txt">
+            <filterchain>
```
Here we have defined patterns for allowed licenses

michaelsembwever replied:
nice.

we should add a comment pointing to https://apache.org/legal/resolved

jacek-lewandowski commented on .build/build-owasp.xml:
```diff
@@ -83,4 +85,67 @@
+            </filterchain>
+        </loadfile>
+
+        <echo message="${wrongly-licensed}" output="${build.dir}/owasp/license-report.txt"/>
```
`build/owasp/license-report.txt` contains list of those jars whose licenses are unknown or are not approved (there have not been filtered out by the above regex filters). So this is the file the developer should look at when analysis fails.

jacek-lewandowski commented on .build/build-owasp.xml:
```diff
@@ -83,4 +85,67 @@
+            </filterchain>
+        </loadfile>
+
+        <loadfile property="license-exceptions-comparable" srcfile="${build.helpers.dir}/license-exceptions.txt" failonerror="false">
```
In `.build/license-exceptions.txt` there is a list of jars we allow despite their license is unknown or not approved. In the end we compare the list of reported jars (see above) and the list of exceptions defined here. If those lists do not match, we fail the build.

michaelsembwever commented on .build/build-git.xml:
```diff
@@ -51,7 +51,34 @@
+        </exec>
+        <filelist id="git.diff.deps-xmls.list" files="${git.diff.deps-xmls}" description="List of changed production sources between HEAD and the latest common ancestor of HEAD and trunk"/>
+        <pathconvert property="git.diff.deps-xmls.path" pathsep="," refid="git.diff.deps-xmls.list">
+        </pathconvert>
```
why? can this go into a private ant target with a descriptive name pls.

jacek-lewandowski replied:
what exactly are you referring to?

michaelsembwever replied:
what does this addition of code/xml do? 

a private ant target with a descriptive name beats a comment.

smiklosovic commented on .build/build-owasp.xml:
```diff
@@ -83,4 +85,67 @@
     </target>
+
+    <target depends="init" name="license-report">
+        <exec executable="jq" osfamily="unix" dir="${basedir}" logError="true" failonerror="false" failifexecutionfails="false">
```
@jacek-lewandowski  @michaelsembwever  This is suspicious. So we need to have `jq` locally, correct? I think we need be very picky and explicit about this. Until now, all you need is git, ant, python and shell.

smiklosovic replied:
we might have a check for this in `build.xml` as we have for Ant.

    <fail message="You need to use Ant of version at least 1.10 to continue.">
      <condition>
        <not>
          <antversion atleast="1.10"/>
        </not>
      </condition>
    </fail>

Even "antversion" is in-built target (I think so), something similar should be done for all other needed tooling and error out as soon as possible if it is not available so it will not fail unnecessarily somewhere down the road.

smiklosovic replied:
I see there is `failifexecutionfails="false", could you explain the consequences? So we are OK when this fails? So if that command is not present, we will just silently skip it?

I think if it just prints some report from owasp, that might be acceptable and we do not need to do anything after all.

jacek-lewandowski replied:
ok, I realized the same and refactored it to run with JavaScript (no worries about j17). Now it seems to be better / cleaner from the user pov.

jacek-lewandowski commented:
@michaelsembwever @smiklosovic - I've refactored it, I agree `jq` may not be a common thing people have in their systems. I don't think it is a big problem - there are some prerequisites to work with the projects and this is not very problematic or uncommon tool.

However, I found implementing it with JavaScript is better. Now we have:
- `allowed-licenses.txt` which includes patterns to match the licenses we approve, it can contain line comments as well
- `license-overrides.txt` which explicitly sets the licenses for jars for which they couldn't be recognized automatically

`check` runs `dependency-check` only if dependencies were touched

I needed to explicitly include Nashorn because JDK-17 does not include it anymore. I know it is GPL but we do not ship it - it is included only in build dependencies, not the `lib` directory.

smiklosovic commented on .build/parent-pom-template.xml:
```diff
@@ -1049,6 +1049,11 @@
         <version>2.3.0</version>
       </dependency>
+      <dependency>
+        <groupId>org.openjdk.nashorn</groupId>
```
@jacek-lewandowski  can we really just add it like that?

jacek-lewandowski replied:
this is only the dependency management - we declare dependency versions here. The dependency in fact is defined in [build-resolver.xml](https://github.com/apache/cassandra/pull/2434/files#diff-b7d8cc9a23d28a1713d5242f83da30838851d0685528abdb937afe7d9b899de6R137-R146)

smiklosovic commented on .build/build-resolver.xml:
```diff
@@ -134,6 +134,16 @@
             <path refid="checkstyle.classpath" classpath="runtime"/>
         </resolve>
+        <resolve>
+<!-- Required for running JavaScript regardless of JDK version, this is only a build dependency, not inlcuded in the distribution -->
```
@jacek-lewandowski could you aligned this? just above `<dependencies>`

jacek-lewandowski replied:
will do

jacek-lewandowski committed:
Run checks in a separate task and fix build warnings

Patch by Jacek Lewandowski; reviewed by Mick Semb Wever and Stefan Miklosovic for CASSANDRA-18618

jacek-lewandowski committed:
DO NOT MERGE - CircleCI config

jacek-lewandowski committed:
DO NOT MERGE - CircleCI config
